### PR TITLE
Implement new indicator - ZLEMA - Zero Lag Exponential Moving Average

### LINKED ISSUE #8118: Implement Zero Lag Exponential Moving Average
#### Expected Behavior
Supports Zero Lag Exponential Moving Average (see [talipp implementation](https://github.com/nardew/talipp/blob/main/talipp/indicators/ZLEMA.py))

Refer to [LEAN Engine > Contributions > Indicators](https://www.quantconnect.com/docs/v2/lean-engine/contributions/indicators).

#### Checklist
- [x] I have completely filled out this template
- [x] I have confirmed that this issue exists on the current `master` branch
- [x] I have confirmed that this is not a duplicate issue by searching [issues](https://github.com/QuantConnect/Lean/issues)
- [x] I have provided detailed steps to reproduce the issue

### PR DESCRIPTION
**Pull request by femtotrader**
#### Description

Implement new indicator - ZLEMA - Zero Lag Exponential Moving Average

#### Related Issue
Closes #8118

#### Motivation and Context

The Zero Lag Exponential Moving Average (ZLEMA) indicator was developed by John Ehlers and Ric Way. The objective is to eliminate the inherent lag associated with all trend following indicators such as averages.

#### Requires Documentation Change

Doc for this indicator should be created

#### How Has This Been Tested?

comparison with talipp computation of ZLEMA(5) of SPY close prices

#### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Refactor (non-breaking change which improves implementation)
- [ ] Performance (non-breaking change which improves performance. Please add associated performance test and results)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Non-functional change (xml comments/documentation/etc)

#### Checklist:

- [x] My code follows the code style of this project.
- [x] I have read the **CONTRIBUTING** [document](https://github.com/QuantConnect/Lean/blob/master/CONTRIBUTING.md).
- [x] I have added tests to cover my changes. 
- [x] All new and existing tests passed.
- [x] My branch follows the naming convention `bug-<issue#>-<description>` or `feature-<issue#>-<description>`

### CODE REVIEW

File: Tests/Indicators/ZeroLagExponentialMovingAverageTests.cs [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+{
+    [TestFixture]
+    public class ZeroLagExponentialMovingAverageTests : CommonIndicatorTests<IndicatorDataPoint>
+    {
+        protected override IndicatorBase<IndicatorDataPoint> CreateIndicator()
+        {
+            return new ZeroLagExponentialMovingAverage(5);
+        }
+
+        protected override string TestFileName => "spy_with_zlema.csv";
```
Comment (AlexCatarino) [inline Tests/Indicators/ZeroLagExponentialMovingAverageTests.cs:32]: Could you please provide the code used to generate this file?
Please check out the following: https://github.com/QuantConnect/Lean/issues/4166#issuecomment-1777061450

File: Tests/TestData/generate_reference_data_from_talipp.py [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+from pathlib import Path
+import talipp
+from talipp.indicators import ZLEMA, McGinleyDynamic  # SISO
+from talipp.ohlcv import OHLCVFactory
+from talipp.indicators import VWMA, ChaikinOsc, CHOP, ForceIndex, IBS, KVO, SOBV  # MISO
+from talipp.indicators import StochRSI, KST  # SIMO
+from talipp.indicators import ChandeKrollStop, SFX, TTM, VTX, ZigZag  # MIMO
+
+
+def main():
```
Comment (Martin-Molinero) [inline Tests/TestData/generate_reference_data_from_talipp.py:12]: Nice!

**Review by Martin-Molinero — CHANGES_REQUESTED** [review]
Hey @femtotrader!
Welcome to Lean & nice work 🙌, leaving some requests and suggestions. Believe many of them apply to the other indicators PRs please take them into consideration and address given this is the first review being done👍

File: Indicators/ZeroLagExponentialMovingAverage.cs [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+    /// </summary>
+    public class ZeroLagExponentialMovingAverage : WindowIndicator<IndicatorDataPoint>, IIndicatorWarmUpPeriodProvider
+    {
+        /// <summary>
+        /// An exponential moving average is used
+        /// </summary>
+        private readonly int _period;
+        private readonly ExponentialMovingAverage _ema;
+        private readonly int _lag;
+        private readonly RollingWindow<IndicatorDataPoint> _rolling_window;
```
[the remark this answers is not in the document]
Comment (Martin-Molinero) [reply Indicators/ZeroLagExponentialMovingAverage.cs:35]: Or actually a better name would be something like `_delayedPrice` being a Delay indicator?

File: Tests/TestData/generate_reference_data_from_talipp.py (hunk 2 above)
Comment (femtotrader) [reply Tests/TestData/generate_reference_data_from_talipp.py:12]: I'm more a Python guy than C#

File: Tests/Indicators/ZeroLagExponentialMovingAverageTests.cs [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+            foreach (var data in TestHelper.GetDataStream(4))
+            {
+                zlema.Update(data);
+            }
+            Assert.IsTrue(zlema.IsReady);
+
+            zlema.Reset();
+
+            TestHelper.AssertIndicatorIsInDefaultState(zlema);
+            //TestHelper.AssertIndicatorIsInDefaultState(zlema._ema);
```
[the remark this answers is not in the document]
Comment (femtotrader) [reply Tests/Indicators/ZeroLagExponentialMovingAverageTests.cs]: Sorry but I didn't understood your last comment here

[the remark this answers is not in the document]
Comment (Martin-Molinero) [reply Tests/Indicators/ZeroLagExponentialMovingAverageTests.cs]: Just replacing `DateTime.UtcNow`  for specific date time instances `new DateTime(2024, 07, 05, X, Y, Z)`, believe makes it more deterministic & performant

File: Indicators/ZeroLagExponentialMovingAverage.cs [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+using System;
+using System.Linq;
+using Fasterflect;
+using MathNet.Numerics;
+using static QuantConnect.Messages;
+
+namespace QuantConnect.Indicators
+{
+    /// <summary>
+    /// Represents the zero lag moving average indicator (ZLEMA)
```
[the remark this answers is not in the document]
Comment (femtotrader) [reply Indicators/ZeroLagExponentialMovingAverage.cs:21]: Done

File: Indicators/ZeroLagExponentialMovingAverage.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (femtotrader) [reply Indicators/ZeroLagExponentialMovingAverage.cs]: Done

File: Indicators/ZeroLagExponentialMovingAverage.cs [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+        /// <param name="input">The input value to this indicator on this time step</param>
+        /// <returns>A new value for this indicator</returns>
+        protected override decimal ComputeNextValue(IReadOnlyWindow<IndicatorDataPoint> window, IndicatorDataPoint input)
+        {
+            _rolling_window.Add(input);
+
+            if (Samples >= _lag + 1)
+            {
+                _ema.Update(input.Time, input.Value + (input.Value - _rolling_window[_lag].Value));
+            } else {
```
[the remark this answers is not in the document]
Comment (femtotrader) [reply Indicators/ZeroLagExponentialMovingAverage.cs]: Done

File: Indicators/ZeroLagExponentialMovingAverage.cs [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+            } else
+            {
+                return 0;
+            }
+        }
+
+        /// <summary>
+        /// Required period, in data points, for the indicator to be ready and fully initialized.
+        /// </summary>
+        public int WarmUpPeriod => _period + (int)Math.Floor(((float)_period) / 2);
```
[the remark this answers is not in the document]
Comment (femtotrader) [reply Indicators/ZeroLagExponentialMovingAverage.cs:95]: I don't see where I should use override. When I set this property as private tests are failing. 

When I use

```
        private override int WarmUpPeriod
        {
            get { return _period + (int)Math.Floor(((float)_period) / 2); }
        }
```

compiler is complaining like so


```
Severity	Code	Description	Project	File	Line	Suppression State
Error (active)	CS0621	'ZeroLagExponentialMovingAverage.WarmUpPeriod': virtual or abstract members cannot be private	QuantConnect.Indicators	C:\Users\Admin\git\github\femto\Lean\Indicators\ZeroLagExponentialMovingAverage.cs	96	
Error (active)	CS0507	'ZeroLagExponentialMovingAverage.WarmUpPeriod': cannot change access modifiers when overriding 'public' inherited member 'WindowIndicator<IndicatorDataPoint>.WarmUpPeriod'	QuantConnect.Indicators	C:\Users\Admin\git\github\femto\Lean\Indicators\ZeroLagExponentialMovingAverage.cs	96	
```

File: Indicators/ZeroLagExponentialMovingAverage.cs [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+        /// </summary>
+        /// <param name="name">The name of this indicator</param>
+        /// <param name="period">The period of the ZLEMA</param>
+        public ZeroLagExponentialMovingAverage(string name, int period)
+            : base(name, period)
+        {
+            _period = period;
+            _lag = (int)Math.Round((period - 1) / 2.0);
+            _ema = new ExponentialMovingAverage(name + "_EMA", period);
+            _rolling_window = new RollingWindow<IndicatorDataPoint>(_lag + 1);
```
[the remark this answers is not in the document]
Comment (femtotrader) [reply Indicators/ZeroLagExponentialMovingAverage.cs]: Done

File: Indicators/ZeroLagExponentialMovingAverage.cs [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+        /// Computes the next value for this indicator from the given state.
+        /// </summary>
+        /// <param name="window">The window of data held in this indicator</param>
+        /// <param name="input">The input value to this indicator on this time step</param>
+        /// <returns>A new value for this indicator</returns>
+        protected override decimal ComputeNextValue(IReadOnlyWindow<IndicatorDataPoint> window, IndicatorDataPoint input)
+        {
+            _rolling_window.Add(input);
+
+            if (Samples >= _lag + 1)
```
[the remark this answers is not in the document]
Comment (femtotrader) [reply Indicators/ZeroLagExponentialMovingAverage.cs]: Done

File: Indicators/ZeroLagExponentialMovingAverage.cs [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+*/
+
+using System;
+using System.Linq;
+using Fasterflect;
+using MathNet.Numerics;
+using static QuantConnect.Messages;
```
[the remark this answers is not in the document]
Comment (femtotrader) [reply Indicators/ZeroLagExponentialMovingAverage.cs]: They have been removed

**Comment by AlexCatarino** [comment]
This PR didn't add `MovingAverageType.Zlema` to [MovingAverageType.cs](https://github.com/QuantConnect/Lean/blob/master/Indicators/MovingAverageType.cs#L21) and [MovingAverageTypeExtensions.cs](https://github.com/QuantConnect/Lean/blob/master/Indicators/MovingAverageTypeExtensions.cs#L23).

**Comment by femtotrader** [comment]
Good catch @AlexCatarino 
I'm going to sleep. Maybe @Martin-Molinero can tackle this.
Sorry for forgeting this

**Pull request merged**